packages/flutter/test/widgets/widget_inspector_test.dart

```diff
@@ -1202,252 +1202,685 @@ class _TestWidgetInspectorService extends TestWidgetInspectorService {
+
+            final Map<String, Object?> jsonObject =
+                json.decode(service.getSelectedWidget(null, 'my-group'))
+                    as Map<String, Object?>;
+            expect(jsonObject, isNot(contains('createdByLocalProject')));
+            final Map<String, Object?> creationLocation =
+                jsonObject['creationLocation']! as Map<String, Object?>;
+            expect(creationLocation, isNotNull);
+            // This RichText widget is created by the build method of the Text widget
+            // thus the creation location is in text.dart not basic.dart
+            final List<String> pathSegmentsFramework =
+                Uri.parse(creationLocation['file']! as String).pathSegments;
+            expect(
+              pathSegmentsFramework.join('/'),
+              endsWith('/flutter/lib/src/widgets/text.dart'),
+            );
+
+            // Strip off /src/widgets/text.dart.
+            final String pubRootFramework =
+                '/${pathSegmentsFramework.take(pathSegmentsFramework.length - 3).join('/')}';
+            service.resetPubRootDirectories(<String>[]);
+            service.addPubRootDirectories(<String>[pubRootFramework]);
+            expect(
+              json.decode(service.getSelectedWidget(null, 'my-group')),
+              contains('createdByLocalProject'),
+            );
+            service.setSelection(elementA, 'my-group');
+            expect(
+              json.decode(service.getSelectedWidget(null, 'my-group')),
+              isNot(contains('createdByLocalProject')),
+            );
+
+            service.resetPubRootDirectories(<String>[]);
+            service
+                .addPubRootDirectories(<String>[pubRootFramework, pubRootTest]);
+            service.setSelection(elementA, 'my-group');
+            expect(
+              json.decode(service.getSelectedWidget(null, 'my-group')),
+              contains('createdByLocalProject'),
+            );
+            service.setSelection(richText, 'my-group');
+            expect(
+              json.decode(service.getSelectedWidget(null, 'my-group')),
+              contains('createdByLocalProject'),
+            );
+          },
+        );
+      });
 
-          testWidgets(
-            'has createdByLocalProject when the element is part of the pubRootDirectory',
-            (WidgetTester tester) async {
-              final Widget widget = Directionality(
-                textDirection: TextDirection.ltr,
-                child: Stack(
-                  children: const <Widget>[
-                    Text('a'),
-                    Text('b', textDirection: TextDirection.ltr),
-                    Text('c', textDirection: TextDirection.ltr),
-                  ],
-                ),
-              );
-              await tester.pumpWidget(widget);
-              final Element elementA = find.text('a').evaluate().first;
+      group('removePubRootDirectories', () {
```

Comment:
add a getting function that returns a copy of the pubrootdirectories, and make it visible only for testing.
Use that to properly unit test the behaviour of add and remove.
Then this can be shaved down to just test the caching and `createdByLocalProject` aspects of the integration.
 Modify the original pubrootdirectories test to use add and remove, instead instead of set. This will satisfy the above, reminder: it is ok for setPubroot directories to go "untested" since it is being deprecated and technicaally will still be tested